### PR TITLE
Support approx aggregation join query

### PR DESCRIPTION
**Pull request by ttt-77**
- [x] Support approx aggregation JOIN query with uniform sampling.
- [x] Add an HTTP inference service specifically for JOIN queries. Since the HTTP inference service yield a single output for each input, the new service significantly speeds up the process.

### CODE REVIEW

**Comment by ttt-77** [comment]
@ddkang Could you please review this PR?

File: aidb/engine/approx_aggregate_join_engine.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+import sqlglot.expressions as exp
+from sqlalchemy.sql import text
+
+from aidb.engine.approx_aggregate_engine import ApproximateAggregateEngine
+from aidb.query.query import Query
+from aidb.utils.constants import (MASS_COL_NAME, NUM_ITEMS_COL_NAME,
+                                  WEIGHT_COL_NAME)
+from aidb.utils.logger import logger
+
+_NUM_PILOT_SAMPLES = 3000000
```
Comment (ddkang) [inline aidb/engine/approx_aggregate_join_engine.py:14]: Are we using a pilot sample size of 3 million?

File: aidb/engine/approx_aggregate_join_engine.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
+from typing import Dict, List
```
Comment (ddkang) [inline aidb/engine/approx_aggregate_join_engine.py:1]: Do we save the outputs of the join conditions in the cache?

File: aidb/engine/approx_aggregate_join_engine.py (hunk 1 above)
Comment (ttt-77) [reply aidb/engine/approx_aggregate_join_engine.py:14]: For a pilot sample size of 1 million, approximately 46 million samples are needed to achieve the desired results. When the size increases to 2 million, the required number decreases to 37 million. Further increasing the size to 3 million reduces the required number to 21 million.

File: aidb/engine/approx_aggregate_join_engine.py (hunk 2 above)
Comment (ttt-77) [reply aidb/engine/approx_aggregate_join_engine.py:1]: Currently, I directly use the HttpInferenceService to get outputs within a User Defined Function. I will modify it by switching to the use of BoundInferenceService.

**Comment by ttt-77** [comment]
@ddkang Could you please review the latest commits?

File: aidb/engine/engine.py [hunk 3]
```diff
@@ -1,30 +1,33 @@
...
       # check validity of user defined function query
+
       if parsed_query.is_udf_query:
         parsed_query.check_udf_query_validity()
       result = None
 
       for parsed_single_query, _ in all_queries:
         if parsed_single_query.is_approx_agg_query:
-          result = asyncio_run(self.execute_aggregate_query(parsed_single_query, **kwargs))
+          if parsed_single_query.is_aqp_join_query:
```
Comment (ddkang) [inline aidb/engine/engine.py:27]: Can we add parsing tests for queries with joins that are not approximate?

File: aidb/query/query.py [hunk 4]
```diff
@@ -400,6 +400,8 @@ def _remove_alias_in_expression(original_list):
           node.set('table', exp.Identifier(this=table_name, quoted=False))
 
         elif isinstance(node.args['this'], exp.Star):
+          if isinstance(parent, exp.AggFunc):
```
Comment (ddkang) [inline aidb/query/query.py:403]: What exactly is this for?

Comment (ttt-77) [reply aidb/query/query.py:403]: Count(*)

File: aidb/inference/bound_inference_service.py [hunk 5]
```diff
@@ -193,19 +194,21 @@ async def _get_inputs_not_in_cache_table(self, inputs: pd.DataFrame, conn):
...
+    in_cache_df_primary = inputs[is_in_cache][normalized_cache_cols]
+    out_cache_df = inputs[~is_in_cache]
+    out_cache_df_primary = out_cache_df[normalized_cache_cols]
 
-    return out_cache_df, out_cache_df_primary
+    return out_cache_df, out_cache_df_primary, in_cache_df_primary
 
 
-  async def infer(self, inputs: pd.DataFrame):
+  async def infer(self, inputs: pd.DataFrame, if_return=False):
```
Comment (ddkang) [inline aidb/inference/bound_inference_service.py:211]: this name is awkward, what does `if_return` mean?

File: aidb/inference/bound_inference_service.py [hunk 6]
```diff
@@ -226,6 +229,40 @@ async def infer(self, inputs: pd.DataFrame):
       await self._insert_in_cache_table(inputs_not_in_cache_primary_cols, conn)
       await self._insert_output_results_in_tables(records_to_insert_in_table, inputs_not_in_cache, conn)
 
+    if if_return:
+      return_res = records_to_insert_in_table.copy()
+
+      # Retrieve cached results
+      tuples_for_condition = []
```
Comment (ddkang) [inline aidb/inference/bound_inference_service.py:236]: what does `tuples_for_condition` mean?

Comment (ttt-77) [reply aidb/inference/bound_inference_service.py]: Store the keys associated with each sampled blob. Rename it to 'sampled_key_list'

File: aidb/inference/bound_inference_service.py (hunk 5 above)
Comment (ttt-77) [reply aidb/inference/bound_inference_service.py]: Renamed it to 'return_inference_results', which means if the 'infer' function should return the inference results.

File: aidb/engine/approx_aggregate_join_engine.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+                            for df_a, df_b in zip(res_store_list, agg_res_list)]
+
+    return res_store_list
+
+
+  def get_additional_required_num_samples_join(
+      self,
+      query: Query,
+      sample_results: List[pd.DataFrame],
+      agg_type_with_cols,
```
Comment (ddkang) [inline aidb/engine/approx_aggregate_join_engine.py:170]: type this

**Review by ddkang — APPROVED** [review]

**Pull request merged**